runtime/tests/interpreter/account_test.go

```diff
@@ -2597,3 +2588,318 @@ func TestInterpretAccount_iteration(t *testing.T) {
+			} else {
+				require.NoError(t, err)
+			}
+		})
+
+		t.Run(fmt.Sprintf("link, continue: %t", continueAfterMutation), func(t *testing.T) {
+			t.Parallel()
+			address := interpreter.NewUnmeteredAddressValueFromBytes([]byte{42})
+
+			inter, _ := testAccount(
+				t,
+				address,
+				true,
+				fmt.Sprintf(`fun test() {
+					account.save(1, to: /storage/foo1)
+					account.save("", to: /storage/foo2)
+					account.link<&Int>(/public/foo1, target: /storage/foo1)
+					account.link<&String>(/public/foo2, target: /storage/foo2)
+
+					account.forEachPublic(fun (path: PublicPath, type: Type): Bool {
+						if type == Type<Capability<&String>>() {
+							account.link<&Int>(/public/foo3, target: /storage/foo1)
+							return %t
+						}
+						return true
+					})
+				}`, continueAfterMutation),
+			)
+
+			_, err := inter.Invoke("test")
+			if continueAfterMutation {
+				require.Error(t, err)
+				require.ErrorAs(t, err, &interpreter.StorageMutatedDuringIterationError{})
+			} else {
+				require.NoError(t, err)
+			}
+		})
+
+		t.Run(fmt.Sprintf("unlink, continue: %t", continueAfterMutation), func(t *testing.T) {
+			t.Parallel()
+			address := interpreter.NewUnmeteredAddressValueFromBytes([]byte{42})
+
+			inter, _ := testAccount(
+				t,
+				address,
+				true,
+				fmt.Sprintf(`fun test() {
+					account.save(1, to: /storage/foo1)
+					account.save("", to: /storage/foo2)
+					account.link<&Int>(/public/foo1, target: /storage/foo1)
+					account.link<&String>(/public/foo2, target: /storage/foo2)
+
+					account.forEachPublic(fun (path: PublicPath, type: Type): Bool {
+						if type == Type<Capability<&String>>() {
+							account.unlink(/public/foo1)
+							return %t
+						}
+						return true
+					})
+				}`, continueAfterMutation),
+			)
+
+			_, err := inter.Invoke("test")
+			if continueAfterMutation {
+				require.Error(t, err)
+				require.ErrorAs(t, err, &interpreter.StorageMutatedDuringIterationError{})
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+
+	test(true)
+	test(false)
+
+	t.Run("state properly cleared on iteration end", func(t *testing.T) {
+		t.Parallel()
+		address := interpreter.NewUnmeteredAddressValueFromBytes([]byte{42})
+
+		inter, _ := testAccount(
+			t,
+			address,
+			true,
+			`fun test() {
+				account.save(1, to: /storage/foo1)
+				account.save(2, to: /storage/foo2)
+				account.save(3, to: /storage/foo3)
+				account.save("qux", to: /storage/foo4)
+
+				account.forEachStored(fun (path: StoragePath, type: Type): Bool {
+					return true
+				})
+				account.save("bar", to: /storage/foo5)
+
+				account.forEachStored(fun (path: StoragePath, type: Type): Bool {
+					account.forEachStored(fun (path: StoragePath, type: Type): Bool {
+						return true
+					})
+					return true
+				})
+				account.save("baz", to: /storage/foo6)
+			}`,
+		)
+
+		_, err := inter.Invoke("test")
+		require.NoError(t, err)
+	})
+}
```

Comment:
Could you please add a test case where an imported function modifies storage? That tests the code for the sub-interpreter